### PR TITLE
Fix for set_interface_alt_setting()

### PR DESCRIPTION
**Pull request by macfreek**
It should no longer return a pipe stalled error if the interface is successfully reset.

### CODE REVIEW

**Comment by macfreek** [comment]
This is a fix for #838.

**Comment by mcuee** [comment]
Maybe it is good to issue a warning before issuing the reset to warn that SetAlternateInterface is failing.
```
  kresult = (*(cInterface->interface))->SetAlternateInterface (cInterface->interface, altsetting);
  if (kresult != kIOReturnSuccess)
    darwin_reset_device (dev_handle);
```

**Comment by macfreek** [comment]
> Maybe it is good to issue a warning before issuing the reset to warn that SetAlternateInterface is failing.

Fully agree. It now prints:
```
...
[ 1.090313] [002c0415] libusb: debug [libusb_set_interface_alt_setting] interface 0 altsetting 0
[ 1.091467] [002c0415] libusb: warning [darwin_set_interface_altsetting] SetAlternateInterface: pipe is stalled
[ 1.138173] [002c0415] libusb: debug [darwin_reset_device] darwin/reset_device: waiting for re-enumeration to complete...
...
```

**Comment by hjelmn** [comment]
Can you squash these into a single commit? Once that is done I will merge this.

File: libusb/os/darwin_usb.c [hunk 1]
```diff
@@ -1568,6 +1568,7 @@ static int darwin_set_interface_altsetting(struct libusb_device_handle *dev_hand
 
   kresult = (*(cInterface->interface))->SetAlternateInterface (cInterface->interface, altsetting);
   if (kresult != kIOReturnSuccess)
+    usbi_warn (HANDLE_CTX (dev_handle), "SetAlternateInterface: %s", darwin_error_str(kresult));
     darwin_reset_device (dev_handle);
```
Comment (tormodvolden) [inline libusb/os/darwin_usb.c:1572]: Missing braces so only the warning is conditional, while the reset is always done.

Comment (macfreek) [reply libusb/os/darwin_usb.c:1572]: Bugger. I clearly have been exposed to Python recently :),.Thanks for the catch!

Comment (macfreek) [reply libusb/os/darwin_usb.c:1572]: Fixed in 6d7069f

**Comment by tormodvolden** [comment]
I would have added to the commit message a reference to commit 065e586b which inadvertently changed this that you are now changing back.

**Comment by tormodvolden** [comment]
If I am allowed to nitpick further, we are maybe not "fixing" the pipe error but ignoring it again, which from what I posted on the bug ticket is also done by the Linux kernel (at least in some cases), no idea what Windows does. And successfully has 3 s'es.

**Comment by mcuee** [comment]
Good point. Typo needs to be fixed. And indeed we are not fixing the pipe error but rather ignore the error. Still the code does try to recover from the pipe error by reseting the device.

**Comment by macfreek** [comment]
Yes, @tormodvolden, you _are_ a nitpicker. And I appreciate it. Thanks! :)

To avoid keeping to push, is the following fine. I'm getting verbose now :)

> MacOS: Silence pipe error in set_interface_alt_setting()
> 
> darwin_set_interface_altsetting() no longer returns the status of the underlying SetAlternateInterface(), but instead the result of a subsequent GetNumEndpoints() call. This reverts to the behaviour prior to commit 065e586.
> 
> Since darwin_set_interface_altsetting() resets the interface after SetAlternateInterface(), one of the consequences is that if SetAlternateInterface() returns a kIOUSBPipeStalled error, and the interface is successfully reset, darwin_set_interface_altsetting() now returns LIBUSB_SUCCESS instead of LIBUSB_ERROR_PIPE.
> 
> Closes: #838
> 
> Signed-off-by: Freek Dijkstra

I think that a comment in the code is warranted too, perhaps both that 
* `darwin_set_interface_altsetting()` returns the status of `GetNumEndpoints()`, not of `SetAlternateInterface()`.
* That the reset of the interface is macOS specific.

Unfortunately, I have no other reasons to specify other than "that's what happened before". Suggestions?

**Comment by tormodvolden** [comment]
One reason is that it is closer to what happens on other platforms like Linux where the kernel hides the PIPE error (but resets all endpoints) if there is only one alternate setting in the interface.

**Comment by macfreek** [comment]
I plan to add the following comment to the code. Please check if I got the terminology correct:

>  /* USB 2.0 specs 9.4.10: If a device only supports a default setting for the specified interface, 
>      then a STALL (kIOUSBPipeStalled) may be returned. Mimick behaviour on other platforms, 
>      e.g. Linux kernel: in this case reset all endpoints, and hide errors. Current implementation
>      resets entire device, instead of single interface, due to historic reasons. */

**Comment by hjelmn** [comment]
@macfreek Looks good. Please add the comment to the commit and I will do the merge.

**Comment by hjelmn** [comment]
Closed via bf943bc

**Comment by macfreek** [comment]
@hjelmn Thanks for the merge, but above all your great care and attention. (And everyone else thanks too!)

(The last push was just after the merge, but no worries - it was literally a space missing after a full stop in the comments. After all the nitpicking, I couldn't let that one slip!)

**Pull request closed without merging**